pkg/issuercerts/issuercerts.go

```diff
@@ -133,25 +134,71 @@ func CheckExpiringSoon(cert *x509.Certificate) error {
 	return nil
 }
 
-// CheckCertAlgoRequirements ensures the certificate respects with the constraints
-// we have posed on the public key and signature algorithms
-func CheckCertAlgoRequirements(cert *x509.Certificate) error {
+// CheckIssuerCertAlgoRequirements ensures the certificate respects with the constraints
+// we have posed on the public key and signature algorithms. Issuer certificates can only
+// be signed by an ECDSA certificate.
+func CheckIssuerCertAlgoRequirements(cert *x509.Certificate) error {
 	if cert.PublicKeyAlgorithm == x509.ECDSA {
-		// this is a safe cast here as we know we are using ECDSA
-		k, ok := cert.PublicKey.(*ecdsa.PublicKey)
-		if !ok {
-			return fmt.Errorf("expected ecdsa.PublicKey but got something %v", cert.PublicKey)
+		err := checkECDSACertRequirements(cert)
+		if err != nil {
+			return err
 		}
-		if k.Params().BitSize != 256 {
-			return fmt.Errorf("must use P-256 curve for public key, instead P-%d was used", k.Params().BitSize)
+	} else {
+		return fmt.Errorf("issuer certificate must use ECDSA for public key algorithm, instead %s was used", cert.PublicKeyAlgorithm)
+	}
+
+	return nil
+}
+
+// CheckTrustAnchorAlgoRequirements ensures the certificate respects with the constraints
+// we have posed on the public key and signature algorithms. Trust anchors can be signed by
+// an ECDSA or RSA certificate.
+func CheckTrustAnchorAlgoRequirements(cert *x509.Certificate) error {
+	if cert.PublicKeyAlgorithm == x509.ECDSA {
+		err := checkECDSACertRequirements(cert)
+		if err != nil {
+			return err
+		}
+	} else if cert.PublicKeyAlgorithm == x509.RSA {
+		err := checkRSACertRequirements(cert)
+		if err != nil {
+			return err
 		}
 	} else {
-		return fmt.Errorf("must use ECDSA for public key algorithm, instead %s was used", cert.PublicKeyAlgorithm)
+		return fmt.Errorf("trust anchor must use ECDSA or RSA for public key algorithm, instead %s was used", cert.PublicKeyAlgorithm)
 	}
 
-	if cert.SignatureAlgorithm != x509.ECDSAWithSHA256 {
+	return nil
+}
+
+func checkECDSACertRequirements(cert *x509.Certificate) error {
+	k, ok := cert.PublicKey.(*ecdsa.PublicKey)
+	if !ok {
+		return fmt.Errorf("expected ecdsa.PublicKey but got something %v", cert.PublicKey)
+	}
+	if k.Params().BitSize != 256 {
+		return fmt.Errorf("must use P-256 curve for public key, instead P-%d was used", k.Params().BitSize)
+	}
+	if cert.SignatureAlgorithm != x509.ECDSAWithSHA256 &&
+		cert.SignatureAlgorithm != x509.SHA256WithRSA {
```

Comment:
Regarding this, @alpeb, although we are checking ECDSA certs, if they come from a RSA certificate, their signature algorithm must be a RSA one. That's why it seems mixed ECDSA and RSA here.

Reply:
Makes sense :+1: